lightning-invoice/src/payment.rs

```diff
@@ -0,0 +1,546 @@
+		Self { retry_attempts, ..self }
+	}
+
+	/// Pays the given [`Invoice`], caching it for later use if a retry is needed.
+	pub fn pay_invoice(&self, invoice: &Invoice) -> Result<(), PaymentError> {
+		let payment_hash = PaymentHash(invoice.payment_hash().clone().into_inner());
+		let mut invoice_cache = self.invoice_cache.lock().unwrap();
+		match invoice_cache.entry(payment_hash) {
+			hash_map::Entry::Vacant(entry) => {
+				self.pay_cached_invoice(invoice)?;
+				entry.insert((invoice.clone(), 0));
+				Ok(())
+			},
+			hash_map::Entry::Occupied(_) => Err(PaymentError::Invoice("payment pending")),
+		}
+	}
+
+	fn pay_cached_invoice(&self, invoice: &Invoice) -> Result<(), PaymentError> {
+		let payer = self.payer.node_id();
+		let payee = invoice.recover_payee_pub_key();
+		let payee_features = invoice.features().cloned();
+		let first_hops = self.payer.first_hops();
+		let last_hops = invoice.route_hints();
+		let final_value_msat = invoice.amount_pico_btc()
+			.ok_or(PaymentError::Invoice("amount missing"))? / 10;
+		let final_cltv = invoice.min_final_cltv_expiry() as u32;
+		let route = self.router.find_route(
+			&payer,
+			&payee,
+			payee_features,
+			Some(&first_hops.iter().collect::<Vec<_>>()),
+			&last_hops,
+			final_value_msat,
+			final_cltv,
+		).map_err(|e| PaymentError::Routing(e))?;
+
+		let payment_hash = PaymentHash(invoice.payment_hash().clone().into_inner());
+		let payment_secret = invoice.payment_secret().clone();
+		self.payer.send_payment(&route, payment_hash, &Some(payment_secret))
+			.map_err(|e| PaymentError::Sending(e))
+	}
+
+	/// Removes the [`Invoice`] cached by the given payment hash.
+	///
+	/// Should be called once a payment has failed or succeeded. This is taken care of when
+	/// [`InvoicePayer`] is used as an [`EventHandler`] but can be called independently as well.
+	pub fn remove_cached_invoice(&self, payment_hash: &PaymentHash) {
+		self.invoice_cache.lock().unwrap().remove(payment_hash);
+	}
+}
+
+impl<P: Deref, R, L: Deref, E> EventHandler for InvoicePayer<P, R, L, E>
+where
+	P::Target: Payer,
+	R: Router,
+	L::Target: Logger,
+	E: EventHandler,
+{
+	fn handle_event(&self, event: &Event) {
+		match event {
+			Event::PaymentFailed { payment_hash, rejected_by_dest, .. } => {
+				let mut invoice_cache = self.invoice_cache.lock().unwrap();
+				if let hash_map::Entry::Occupied(mut entry) = invoice_cache.entry(*payment_hash) {
+					let (invoice, attempts) = entry.get_mut();
+					*attempts += 1;
+					if !rejected_by_dest {
+						let max_payment_attempts = self.retry_attempts + 1;
+						if *attempts < max_payment_attempts {
+							if self.pay_cached_invoice(invoice).is_ok() {
+								log_trace!(self.logger, "Payment {} failed; retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
+								return;
+							} else {
+								log_trace!(self.logger, "Error retrying payment {}; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
```

Comment:
I think `entry.remove()` will be called if we hit this line, but it seems like we could still be under the max # payment attempts?